### PR TITLE
Hardening LB SourceRanges filtering for IPVS proxier

### LINKED ISSUE #119656: Can't restrict access from K8s nodes with loadBalancerSourceRanges for proxy-mode=ipvs
### What happened?

When using loadBalancerSourceRanges to allow access only from CIDRs outside the cluster segment, access from the K8s nodes (or pods with hostNetwork:true) should be rejected, but for proxy-mode=ipvs it works.

/sig network
/area kube-proxy
/area ipvs

### What did you expect to happen?

Access from K8s nodes should be rejected in the same way as for proxy-mode=iptables

### How can we reproduce it (as minimally and precisely as possible)?

Define a service with loadBalancerSourceRanges that allows access from a cidr outside the cluster, e.g.

```yaml
apiVersion: v1
kind: Service
metadata:
  name: alpine
spec:
  loadBalancerSourceRanges:
#  - "192.168.1.0/24"
  - "192.168.2.0/24"
#  - "fd00::c0a8:100/120"
  - "fd00::c0a8:200/120"
  ipFamilyPolicy: RequireDualStack
  selector:
    app: alpine
  type: LoadBalancer
  allocateLoadBalancerNodePorts: false
  ports:
  - port: 5001
    name: nc
```
The K8s nodes have addresses on 192.168.1.0/24 and fd00::c0a8:100/120.

Make sure a loadBalancerIP is assigned (in this example 10.0.0.0/32) and try to access it from a K8s node:
```
# With proxy.mode=ipvs
> nc 10.0.0.0:5001
alpine-f6c8bd7db-zljnl    # (works)
# With proxy.mode=iptables
# nc 10.0.0.0:5001
# (hangs...)
```

### Anything else we need to know?

Same for both ipv4 and ipv6.

This is related to https://github.com/kubernetes/kubernetes/issues/119524 since a fix requires that the `nodeID` passed to the `kube-proxy` is correct. Please see https://github.com/kubernetes/kubernetes/issues/119524#issuecomment-1655389139

### Kubernetes version

v1.27.4 and master

### Cloud provider

None

### OS version

Linux 6.4.4

### Install tools

None

### Container runtime (CRI) and version (if applicable)

N/A

### Related plugins (CNI, CSI, ...) and versions (if applicable)

N/A

### PR DESCRIPTION
**Pull request by aroradaman**
#### What type of PR is this?
/kind bug

#### Which issue(s) this PR fixes:

Fixes #119656

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/proxy/ipvs/proxier.go [hunk 1]
```diff
@@ -1637,6 +1648,17 @@ func (proxier *Proxier) writeIptablesRules() {
...
 
+	// disable LB access from node
+	// for IPVS src and dst both would be lbIP
+	for _, lbIP := range proxier.lbNoNodeAccessIPs.UnsortedList() {
+		proxier.filterRules.Write(
+			"-A", string(utiliptables.ChainOutput),
+			"-s", lbIP,
+			"-m", "ipvs", "--vaddr", lbIP,
+			"-j", "DROP",
+		)
```
Comment (aroradaman) [inline pkg/proxy/ipvs/proxier.go:1659]: Tried this locally on kind, every iptables-restore call is creating a duplicate entry in iptable rules. 

After few minutes I ended up with 18 exact same rules 😔 

```
╰─$ docker exec -it kind-worker /bin/bash                                                                                                                                                                       130 ↵
root@kind-worker:/# iptables-save | grep ipvs 
-A INPUT -m comment --comment "kubernetes ipvs access filter" -j KUBE-IPVS-FILTER
-A OUTPUT -s 172.18.255.200/32 -m ipvs --vaddr 172.18.255.200 -j DROP
-A OUTPUT -s 172.18.255.200/32 -m ipvs --vaddr 172.18.255.200 -j DROP
-A OUTPUT -s 172.18.255.200/32 -m ipvs --vaddr 172.18.255.200 -j DROP
-A OUTPUT -s 172.18.255.200/32 -m ipvs --vaddr 172.18.255.200 -j DROP
-A OUTPUT -s 172.18.255.200/32 -m ipvs --vaddr 172.18.255.200 -j DROP
-A OUTPUT -s 172.18.255.200/32 -m ipvs --vaddr 172.18.255.200 -j DROP
-A OUTPUT -s 172.18.255.200/32 -m ipvs --vaddr 172.18.255.200 -j DROP
-A OUTPUT -s 172.18.255.200/32 -m ipvs --vaddr 172.18.255.200 -j DROP
-A OUTPUT -s 172.18.255.200/32 -m ipvs --vaddr 172.18.255.200 -j DROP
-A OUTPUT -s 172.18.255.200/32 -m ipvs --vaddr 172.18.255.200 -j DROP
-A OUTPUT -s 172.18.255.200/32 -m ipvs --vaddr 172.18.255.200 -j DROP
-A OUTPUT -s 172.18.255.200/32 -m ipvs --vaddr 172.18.255.200 -j DROP
-A OUTPUT -s 172.18.255.200/32 -m ipvs --vaddr 172.18.255.200 -j DROP
-A OUTPUT -s 172.18.255.200/32 -m ipvs --vaddr 172.18.255.200 -j DROP
-A OUTPUT -s 172.18.255.200/32 -m ipvs --vaddr 172.18.255.200 -j DROP
-A OUTPUT -s 172.18.255.200/32 -m ipvs --vaddr 172.18.255.200 -j DROP
-A OUTPUT -s 172.18.255.200/32 -m ipvs --vaddr 172.18.255.200 -j DROP
-A OUTPUT -s 172.18.255.200/32 -m ipvs --vaddr 172.18.255.200 -j DROP
root@kind-worker:/# 
```

Also, checked the `proxier.filterRules` it only has single entry for this rule. I guess duplication has something to do with the flags which we pass to restore call.

**Review by uablrek — APPROVED** [review]

File: pkg/proxy/ipvs/proxier.go [hunk 2]
```diff
@@ -1246,6 +1254,9 @@ func (proxier *Proxier) syncProxyRules() {
 						continue
 					}
 					proxier.ipsetList[kubeLoadBalancerSourceIPSet].activeEntries.Insert(entry.String())
+				} else {
```
[the remark this answers is not in the document]
Comment (uablrek) [reply pkg/proxy/ipvs/proxier.go:1264]: Forget the above. The use is different from iptables

**Comment by uablrek** [comment]
I have tested that access from the node is rejected if the nodeIp is _not_ in `loadBalancerSourceRanges`, and that it works if the nodeIP CIDR is in it.

I have not tested with a POD with `hostNetwork:true`, but that should be the same as from a node. To use `hostNetwork:true` makes it possible to write an e2e test for this, but IMO it should not be required in this PR, but perhaps in a separate one.

Here are some iptables dumps for info:
```
# iptables -L
...
Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
KUBE-IPVS-OUT-FILTER  all  --  anywhere             anywhere             /* kubernetes ipvs access filter */
KUBE-FIREWALL  all  --  anywhere             anywhere            
...
Chain KUBE-IPVS-OUT-FILTER (1 references)
target     prot opt source               destination         
DROP       all  --  10.0.0.10            anywhere             vaddr 10.0.0.10
# iptables -S
...
-N KUBE-IPVS-OUT-FILTER
...
-A KUBE-IPVS-OUT-FILTER -s 10.0.0.10/32 -m ipvs --vaddr 10.0.0.10 -j DROP
```

**Review by uablrek — CHANGES_REQUESTED** [review]
Seems like the port must be considered

**Comment by uablrek** [comment]
Sorry, I was too hasty on the "approve" button. The port must be included. Consider this case:
```yaml
kind: Service
metadata:
  name: alpine2
spec:
  ipFamilyPolicy: RequireDualStack
  selector:
    app: alpine2
  type: LoadBalancer
  allocateLoadBalancerNodePorts: false
  ports:
  - port: 8001
    name: nc
```
With loadBalancerIP `10.0.0.10`. The service is accessible from a K8s node on address `10.0.0.10:8001`. But if _another_ service is created:
```yaml
apiVersion: v1
kind: Service
metadata:
  name: alpine1
spec:
  loadBalancerSourceRanges:
  - "192.168.2.0/24"
  - "fd00::c0a8:200/120"
  ipFamilyPolicy: RequireDualStack
  selector:
    app: alpine1
  type: LoadBalancer
  allocateLoadBalancerNodePorts: false
  ports:
  - port: 7001
    name: nc
```
And is _also_ assigned the loadBalancerIP `10.0.0.10`, then the first service is not accessible any more from K8s nodes.

This works as expected with `proxy-mode=iptables`, access from nodes still works on the firsts service even when the second is created with the same loadBalancerIP.

I am not at all sure that this bug is worth solving. Nobody has complained, and the nft proxier is on the way. But if this is not solved the constraint should be documented with something like:
```
For proxy-mode=ipvs the loadBalancerSourceRanges in can not be used to deny access to the service from
K8s nodes or from PODs with "hostNetwork:true".
```

**Comment by uablrek** [comment]
It might be worth introducing a new "ipset". But I don't know...

**Comment by aroradaman** [comment]
> It might be worth introducing a new "ipset". But I don't know...

I initially thought of hash:ip,port,ip set only but I was not sure if could match ipset on ipvs extension.

**Comment by aroradaman** [comment]
@uablrek I updated the rules to match on IP + Port + Protocol.

**Review by uablrek — APPROVED** [review]

**Comment by uablrek** [comment]
IMO a very tidy PR :+1: 

/lgtm

**Comment by aroradaman** [comment]
/hold cancel
I added the hold label for code freeze before https://github.com/kubernetes/kubernetes/pull/119791#issuecomment-1667400344

**Pull request merged**